test: add MAX_DISCONNECTED_TX_POOL_BYTES, chainlimits coverage

instagibbs opened the pull request:
`DisconnectedBlockTransactions::LimitMemoryUsage()` has unit test coverage, but the default value end to end doesn't have coverage.

This test adds exercised coverage of memory limiting of the disconnect pool, and some basic behavior sanity checks.

Another test added is making sure chainlimits are being respected on reorg, and the expected transactions pruned.

Lastly, fix the existing test case which was using a deficient test via directly inducing reorgs with `invalidateblock`

instagibbs committed:
functional test: add MAX_DISCONNECTED_TX_POOL_BYTES coverage

instagibbs commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -97,6 +98,53 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, fee=100_000):
+        wallet.rescan_utxos()
+
+        # Make 200 outputs to spend via wallet
+        wallet.send_self_transfer_multi(from_node=self.nodes[0], num_outputs=200)
```
MiniWallet doesn't appear to support more than 49 utxos at a time? Unclear to me what's going on so I did this instead.

theStack replied:
Not exactly sure if this is what you are asking, but: new blocks with coinbase transaction outputs for MiniWallet can be generated by passing an instance of the latter to `self.generate` (instead of the node object), i.e. in this case:
```
        self.generate(wallet, 300)
```
(added +100 to let the coins mature)

instagibbs replied:
Oh, mine to wallet, not to node which has attached wallet :facepalm:

instagibbs commented:
I don't think I'm actually hitting that particular limit, getting weird results when looking closer, draft for now while I look

instagibbs committed:
test: check chainlimits respects on reorg

instagibbs committed:
test: fix transaction_graph_test reorg test

The current test directly uses invalidatblock to trigger
mempool re-entry of transactions. Unfortunately, the
behavior doesn't match what a real reorg would look like. As
a result you get surprising behavior such as the mempool
descendant chain limits being exceeded, or if a fork is
greater than 10 blocks deep, evicted block transactions stop
being submitted back into in the mempool.

Fix this by preparing an empty fork chain, and then
continuing with the logic, finally submitting the fork chain
once the rest of the test is prepared. This triggers a more
typical codepath.

Also, extend the descendant limit to 100, like ancestor
limit.

mzumsande commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -190,7 +197,7 @@ def test_chainlimits_exceeded(self):
 
     def run_test(self):
         # Use batch size limited by DEFAULT_ANCESTOR_LIMIT = 25 to not fire "too many unconfirmed parents" error.
```
This comment is outdated, because `DEFAULT_ANCESTOR_LIMIT = 25` is no longer after the last commit. Not sure about the reason why it was used before, is it no longer valid?

Also, changing DEFAULT_ANCESTOR_LIMIT to 100 could be mentioned in the commit message, it also affects the other subtests.

instagibbs replied:
Kept the test setup as is, since it tests a meaningful 3 block reorg. Changed the comment because the motivation is different.

mzumsande reviewed (commented):
Concept ACK

instagibbs commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -97,10 +106,102 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, fee=100_000):
+        small_child_txs = []
+        aggregate_serialized_size = 0
+        while aggregate_serialized_size < MAX_DISCONNECTED_TX_POOL_BYTES:
+            # Mine parents in FIFO order via priority (priority map wiped during reorg)
```
oops, yep, removed

mzumsande reviewed (commented):
Code Review ACK 2b202e9db56487e658fc41089178f31ef4259a0d

i-am-yuvi commented:
Concept ACK

yuvicc reviewed (commented):
Concept ACK

yuvicc reviewed (commented):
tACK 2b202e9db56487e658fc41089178f31ef4259a0d

- `DisconnectTip()` & reconnect via `invalidateblock` and `reconsiderblock`
- `LimitMemoryUsage()`: exercised by flooding > 20 MB
-  `test_chainlimits_exceeded`

Logs
```
2025-05-21T16:57:42.023000Z TestFramework (INFO): PRNG seed is: 4085491027579870526
2025-05-21T16:57:42.024000Z TestFramework (INFO): Initializing test directory /var/folders/jb/wlbrz0t95vl58wzxjt75wqmh0000gn/T/bitcoin_func_test_jb27si_k
2025-05-21T16:57:42.337000Z TestFramework (INFO): Creating 100 transactions...
2025-05-21T16:57:42.717000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2025-05-21T16:57:42.722000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2025-05-21T16:57:44.161000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2025-05-21T16:57:44.166000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2025-05-21T16:57:46.466000Z TestFramework (INFO): The batch of 25 transactions has been accepted into the mempool.
2025-05-21T16:57:46.472000Z TestFramework (INFO): All of the transactions from the current batch have been mined into a block.
2025-05-21T16:57:48.423000Z TestFramework (INFO): The last batch of 25 transactions has been accepted into the mempool.
2025-05-21T16:57:48.571000Z TestFramework (INFO): All of the recently mined transactions have been re-added into the mempool in 0.1478879451751709 seconds.
2025-05-21T16:57:48.571000Z TestFramework (INFO): Checking descendants/ancestors properties of all of the in-mempool transactions...
2025-05-21T16:57:48.633000Z TestFramework (INFO): Creating independent transactions to test MAX_DISCONNECTED_TX_POOL_BYTES limit during reorg
2025-05-21T16:57:53.509000Z TestFramework (INFO): Check that too long chains on reorg are handled
2025-05-21T16:57:55.389000Z TestFramework (INFO): Stopping nodes
2025-05-21T16:57:55.494000Z TestFramework (INFO): Cleaning up /var/folders/jb/wlbrz0t95vl58wzxjt75wqmh0000gn/T/bitcoin_func_test_jb27si_k on exit
2025-05-21T16:57:55.494000Z TestFramework (INFO): Tests successful
```

TheCharlatan commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -97,10 +99,67 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, fee=100_000):
+        # Set up empty fork blocks ahead of time, needs to be longer than full fork made later
+        fork_length = 60
+        fork_blocks = self.generate(self.nodes[0], fork_length)
+        self.nodes[0].invalidateblock(fork_blocks[0])
```
Just a comment, I think it is kind of fine to do the reorgs through invalidateblock instead of connecting and disconnecting nodes from each other, but the disconnectpool behaviour here is kind of on the line where I would tend to prefer exercising the actual code through `ActiveBestChainStep`, since the two methods do not share the same object. I think to test the mempool in general it is fine to do, both methods at least use the same object at runtime.

instagibbs replied:
did a proper reorg using blocktools to prep the fork.

sedited reviewed (commented):
Concept ACK

maflcko reviewed (commented):
lgtm ACK 2b202e9db56487e658fc41089178f31ef4259a0d 🛤

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm ACK 2b202e9db56487e658fc41089178f31ef4259a0d 🛤
SBccUIAZKcRyVW63zp8B0dLkfguSYFAfLhxM+uFjRVeT01zjRQ9onvrSXYmRRKWtG0RZDIGQjc6YaDks3w3JDw==
```

</details>

instagibbs commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -155,11 +155,46 @@ def test_max_disconnect_pool_bytes(self):
+    def test_chainlimits_exceeded(self):
+        self.log.info('Check that too long chains on reorg are handled')
+
+        # Generate coins test
```
deleted

instagibbs commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -97,10 +99,67 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, fee=100_000):
+        # but not all, and any time parent is dropped, child is also removed
+        self.nodes[0].reconsiderblock(fork_blocks[0])
+        expected_parent_count = len(large_std_txs) - 2
+        assert_equal(len(self.nodes[0].getrawmempool()), expected_parent_count * 2)
```
done

instagibbs commented on test/functional/mempool_updatefromblock.py:
```diff
@@ -73,17 +84,15 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, fee=100_000):
                 self.log.info('All of the transactions from the current batch have been mined into a block.')
             elif tx_count == size:
-                # At the end all of the mined blocks are invalidated, and all of the created
+                # At the end all of the mined the old fork is revalidated to cause reorg, and all of the created
```
done

sedited reviewed (approved):
Thanks for entertaining my suggestion :)

ACK 84aa484d45e2fb3c1149941ef23779e4adb983d9

yuvicc reviewed (commented):
Post-Merge ACK 84aa484d45e2fb3c1149941ef23779e4adb983d9
